### PR TITLE
Production pull - multiple fixed issues.

### PR DESCRIPTION
**Pull request by bubby932**
* Fixes an issue where creating instances would actually create two, causing an aliasing issue that could lead to problems down the line.
* Fixes an issue where instances would always be persistent regardless of their persistence setting, resulting in a managed memory leak.
* Bumps `firebase` and `firebase-admin` dependencies.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,8 +20,8 @@
     "express-fileupload": "^1.2.1",
     "express-rate-limit": "^6.6.0",
     "express-slow-down": "^1.5.0",
-    "firebase": "^9.10.0",
-    "firebase-admin": "^11.0.1",
+    "firebase": "^9.11.0",
```
Comment (bubby932) [inline package.json:23]: Suggested and approved (+tested) by dependabot & @bubby932 respectively.

File: package.json [hunk 2]
```diff
@@ -20,8 +20,8 @@
...
+    "firebase-admin": "^11.1.0",
```
Comment (bubby932) [inline package.json:24]: Merged and tested by @bubby932.

File: routers/matchmaking.js [hunk 3]
```diff
@@ -144,16 +144,15 @@ class RoomSession {
         clearInterval(this.#eventLoopHandle);
         clearInterval(this.#automaticCleanupHandle);
 
-        console.log(`CLEANUP OF INSTANCE ${this.InstanceId}`);
+        console.log(`INSTANCE ${this.InstanceId} MARKED FOR CLEANUP`);
```
Comment (bubby932) [inline routers/matchmaking.js:147]: Improves phrasing.

File: routers/matchmaking.js [hunk 4]
```diff
@@ -144,16 +144,15 @@ class RoomSession {
...
     }
     AutomaticInstanceCleanup() {
-        // damn i really messed up how TTL should work originally haha
         if(this.Persistent || this.AgeWithoutPlayer < this.TTL) return;
 
         if(typeof this.Players != 'object') this.Players = [];
         if(this.Players.length < 1) {
             this.InstanceCleanup();
-            console.log(`AUTOMATIC CLEANUP OF INSTANCE ${this.InstanceId}`);
+            console.log(`INSTANCE ${this.InstanceId} TERMINATED`);
```
Comment (bubby932) [inline routers/matchmaking.js:155]: Clearer terminology, previously it wasn't clear whether the instance was actually garbage collected or just marked grey.

File: routers/matchmaking.js [hunk 5]
```diff
@@ -194,6 +191,10 @@ var GlobalRoomInstances = Object.create(null);
 
 setInterval(CleanupInstances, 300 * 1000);
 
+if (require('../config.json').debug_trace_instances) {
```
Comment (bubby932) [inline routers/matchmaking.js:194]: Useful debugging tool - logs instance data every 30 seconds if there is at least 1 instance open and the `debug_trace_instances` config variable is truthy.

File: routers/matchmaking.js [hunk 6]
```diff
@@ -217,11 +218,11 @@ async function GetInstances(RoomId = null) {
 }
 
 async function GetInstanceById(RoomId, InstanceId) {
-    return SubroomInstances[RoomId]?.find(item => item.InstanceId === InstanceId);
+    return SubroomInstances[RoomId]?.find(item => item.InstanceId == InstanceId);
```
Comment (bubby932) [inline routers/matchmaking.js:221]: Less confusion on types.

File: routers/matchmaking.js [hunk 7]
```diff
@@ -217,11 +218,11 @@ async function GetInstances(RoomId = null) {
...
 }
 
 async function GetInstanceByJoinCode(RoomId, JoinCode) {
-    return SubroomInstances[RoomId]?.find(item => item.JoinCode === JoinCode);
+    return SubroomInstances[RoomId]?.find(item => item.JoinCode == JoinCode);
```
Comment (bubby932) [inline routers/matchmaking.js:225]: Removing === again.

File: routers/matchmaking.js [hunk 8]
```diff
@@ -231,21 +232,30 @@ async function SetInstances(RoomId, InstanceList) {
 async function SetInstance(RoomId, InstanceId, Instance) {
     if(!Object.keys(SubroomInstances).includes(RoomId)) SubroomInstances[RoomId] = [];
 
-    var index = SubroomInstances[RoomId].findIndex(item => item.InstanceId === InstanceId);
+    var index = SubroomInstances[RoomId].findIndex(item => item.InstanceId == InstanceId);
```
Comment (bubby932) [inline routers/matchmaking.js:235]: And again.

File: routers/matchmaking.js [hunk 9]
```diff
@@ -231,21 +232,30 @@ async function SetInstances(RoomId, InstanceList) {
...
     if(index < 0) SubroomInstances[RoomId].push(Instance);
     else SubroomInstances[RoomId][index] = Instance;
 }
 
 async function CreateInstance(RoomId, SubroomId, MatchmakingMode, TTL, Persistent, MaxPlayers) {
     if(!Object.keys(SubroomInstances).includes(RoomId)) SubroomInstances[RoomId] = [];
 
-    const room = new RoomSession(RoomId, SubroomId, MatchmakingMode, TTL, true, MaxPlayers);
+    const room = new RoomSession(RoomId, SubroomId, MatchmakingMode, TTL, Persistent, MaxPlayers);
```
Comment (bubby932) [inline routers/matchmaking.js:243]: Fixed an issue where instances would always be persistent, this was a temporary fix to another problem where instances would close too early, which turned out to be a logic error, however this stopgap measure was never removed.

File: routers/matchmaking.js [hunk 10]
```diff
@@ -231,21 +232,30 @@ async function SetInstances(RoomId, InstanceList) {
...
     SubroomInstances[RoomId].push(room);
 
     console.log(`New instance of room ${RoomId} created with InstanceId of ${room.InstanceId} and a join code of ${room.JoinCode}`);
     return room;
 }
 
+async function LogInstanceTable() {
```
Comment (bubby932) [inline routers/matchmaking.js:250]: Instance logging function for `debug_trace_instances`

File: routers/matchmaking.js [hunk 11]
```diff
@@ -256,5 +266,6 @@ module.exports = {
     GetInstanceByJoinCode: GetInstanceByJoinCode,
     SetInstances: SetInstances,
     SetInstance: SetInstance,
-    CreateInstance: CreateInstance
+    CreateInstance: CreateInstance,
+    LogInstanceTable: LogInstanceTable
```
Comment (bubby932) [inline routers/matchmaking.js:270]: Added export for logging instances.

**Review by bubby932 — COMMENTED** [review]
Explained every single change with a justification & elaborated more on what was actually done.

**Review by Bobrobot1 — APPROVED** [review]
Looks great. Glad you fixed that.

**Pull request merged**